core/shared/src/main/scala/fs2/text.scala

```diff
@@ -306,11 +310,88 @@ object text {
 
   /** Encodes a stream of `String` in to a stream of bytes using the given charset. */
   def encode[F[_]](charset: Charset): Pipe[F, String, Byte] =
-    _.mapChunks(c => c.flatMap(s => Chunk.array(s.getBytes(charset))))
+    encodeC(charset)(_).unchunks
 
   /** Encodes a stream of `String` in to a stream of `Chunk[Byte]` using the given charset. */
-  def encodeC[F[_]](charset: Charset): Pipe[F, String, Chunk[Byte]] =
-    _.mapChunks(_.map(s => Chunk.array(s.getBytes(charset))))
+  def encodeC[F[_]](charset: Charset): Pipe[F, String, Chunk[Byte]] = {
+    def encodeC(
+        encoder: CharsetEncoder,
+        acc: Chunk[Char],
+        s: Stream[F, Chunk[Char]]
+    ): Pull[F, Chunk[Byte], Unit] =
+      s.pull.uncons1.flatMap { r =>
+        val toEncode = r match {
+          case Some((c, _)) => acc ++ c
+          case None         => acc
+        }
+
+        val isLast = r.isEmpty
+        val outBufferSize = (encoder.maxBytesPerChar() * toEncode.size).ceil.toInt
```

Comment:
```suggestion
        val outBufferSize = math.max(encoder.maxBytesPerChar(), encoder.averageBytesPerChar() * toEncode.size).toInt
```

would be more memory-efficient, especially with UTF-8 (average: 1.1, max: 4). BUT: It breaks the UTF-8 roundtrip test as it can affect chunking if not all chars can directly be written to the output. I'm not sure whether chunking is actually something that needs to be preserved strictly or whether that tests ensures this more by accident?

Reply:
> I'm not sure whether chunking is actually something that needs to be preserved strictly or whether that tests ensures this more by accident?

I don't think so, see this note.

https://github.com/typelevel/fs2/blob/51c976855db1afa73c3ff6f6f919aa131076d8c7/core/shared/src/main/scala/fs2/Stream.scala#L117-L137